design/unnumbered-bgp.md

```diff
@@ -0,0 +1,436 @@
+Docs are unclear if needed.
+
+#### BGP RouterID
+
+//TODO is there a requirement for that?
+
+### Traffic forwarding
+
+FRR advertises IPv4 prefixes with IPv6 next hops using encoding specified in RFC 5549 (now RFC 8950).
+
+### The API Extension
+
+#### frr-k8s side
+
+```go
+...
+
+type Neighbor struct {
+  Description string `json:"description"` // optional
+  Address string `json:"address"` // must be empty if interface != nil
+  Interfaces []string `json:"interface"` //  
+  PeerGroup string `json:"peerGroup"` // optional, otherwise value <interface>-peergroup
+  RemoteAS string `json:"remoteAS"` // it can be "ASN number or internal or external [or auto]"
+  Port uint16 `json:"port,omitempty"`
+  PasswordSecret v1.SecretReference `json:"password,omitempty"`
+  HoldTime metav1.Duration `json:"holdTime,omitempty"`
+  KeepaliveTime metav1.Duration `json:"keepaliveTime,omitempty"`
+  EBGPMultiHop bool `json:"ebgpMultiHop,omitempty"`
+  BFDProfile string `json:"bfdProfile,omitempty"`
+  ToAdvertise Advertise `json:"toAdvertise,omitempty"`
+  ToReceive Receive `json:"toReceive,omitempty"`
+}
+
+...
+```
+
+#### MetalLB
+
+```go
+...
+type BGPPeerSpec struct {
+	MyASN uint32 `json:"myASN"`
+	ASN uint32 `json:"peerASN"` // or internal or external
+	Address string `json:"peerAddress"` // or interface name e.g. net0
+	SrcAddress string `json:"sourceAddress,omitempty"`
+	Port uint16 `json:"peerPort,omitempty"`
+	HoldTime metav1.Duration `json:"holdTime,omitempty"`
+	KeepaliveTime metav1.Duration `json:"keepaliveTime,omitempty"`
+	ConnectTime *metav1.Duration `json:"connectTime,omitempty"`
+	RouterID string `json:"routerID,omitempty"`
+	NodeSelectors []metav1.LabelSelector `json:"nodeSelectors,omitempty"`
+	Password string `json:"password,omitempty"`
+	PasswordSecret v1.SecretReference `json:"passwordSecret,omitempty"`
+	BFDProfile string `json:"bfdProfile,omitempty"`
+	EnableGracefulRestart bool `json:"enableGracefulRestart,omitempty"`
+	EBGPMultiHop bool `json:"ebgpMultiHop,omitempty"`
+	VRFName string `json:"vrf,omitempty"`
+	DisableMP bool `json:"disableMP,omitempty"`
+}
+
+```
+
+Example CR
+
+```yaml
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  name: peer1
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  peerASN: external
+  peerAddress: net0
+```
+
+The resulting configuration will be something like:
+
+```yaml
+bgp:
+  routers:
+    - asn: 64512
+      neighbors:
+      //TODO
+```
+
+
+### Testing
+
+#### E2E tests
+
+- Modify the testing infra to include point-to-point links which means
+connect containers using VETH, and not through a bridge. We need NET-ADMIN cap (or sudo) when creating the VETH
+- Create/Delete peer per specific test not to take resources from runners outside the runtime of the test
+- This peer needs different configuration (not IPs, interface names), keep a raw config
+- There is no straightforward way to test unnumbered in the primary node interface. The testing setup (kind creation)
```

Comment:
I wonder if there's some value in adding another worker+container that are peered using unnumbered (no p2p, but there will be no conflicts as they're the only routers in the network using it) to get some feedback on peering with the primary node interface? wdyt?

Reply:
Do you mean to have a single worker as speaker, and have single external container frr connected on the same docker network bridge? 

My concern is that locally the linux bridge does not forwarding RA when LLA address is the source (at least in my fedora laptop).

For sure we need to verifiy how that works on primary interface, and especially when the interface has also a non-LLA address configured.

Reply:
> My concern is that locally the linux bridge does not forwarding RA when LLA address is the source (at least in my fedora laptop).

oh, didn't you run into the "random" peering we talked about?

> Do you mean to have a single worker as speaker, and have single external container frr connected on the same docker network bridge?

right (adding another worker / designating one), then the "randomness" will be okay - since there are only 2 peers on the same network using unnumbered they'll always peer with each other